flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor;
+import org.apache.flink.table.types.extraction.utils.FunctionResultTemplate;
+import org.apache.flink.table.types.extraction.utils.FunctionSignatureTemplate;
+import org.apache.flink.table.types.inference.InputTypeStrategies;
+import org.apache.flink.table.types.inference.InputTypeStrategy;
+import org.apache.flink.table.types.inference.TypeInference;
+import org.apache.flink.table.types.inference.TypeStrategies;
+import org.apache.flink.table.types.inference.TypeStrategy;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.types.extraction.utils.ExtractionUtils.extractionError;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createGenericResultExtraction;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createParameterAndReturnTypeVerification;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createParameterSignatureExtraction;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createParameterVerification;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createParameterWithAccumulatorVerification;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createParameterWithArgumentVerification;
+import static org.apache.flink.table.types.extraction.utils.FunctionMappingExtractor.createReturnTypeResultExtraction;
+
+/**
+ * Reflection-based utility for extracting a {@link TypeInference} from a supported subclass of
+ * {@link UserDefinedFunction}.
+ *
+ * <p>The behavior of this utility can be influenced by {@link DataTypeHint}s and {@link FunctionHint}s
+ * which have higher precedence than the reflective information.
+ *
+ * <p>Note: This utility assumes that functions have been validated before regarding accessibility of
+ * class/methods and serializability.
+ */
+@Internal
+public final class TypeInferenceExtractor {
+
+	/**
+	 * Extracts a type inference from a {@link ScalarFunction}.
+	 */
+	public static TypeInference forScalarFunction(DataTypeLookup lookup, Class<? extends ScalarFunction> function) {
+		final FunctionMappingExtractor mappingExtractor = new FunctionMappingExtractor(
+			lookup,
+			function,
+			"eval",
+			createParameterSignatureExtraction(0),
+			null,
+			createReturnTypeResultExtraction(),
+			createParameterAndReturnTypeVerification());
+		return extractTypeInference(mappingExtractor);
+	}
+
+	/**
+	 * Extracts a type inference from a {@link AggregateFunction}.
+	 */
+	public static TypeInference forAggregateFunction(DataTypeLookup lookup, Class<? extends AggregateFunction> function) {
```

Comment:
nit: My IDE reports "Raw use of parameterized class AggregateFunction". It does not make a big difference, but just to make the IDE happy we could change it `Class<? extends AggregateFunction<?, ?>>`

The same for subsequent methods.